File: go/problems/single_number.go
package problems

// https://leetcode.com/problems/single-number/
// https://leetcode.com/problems/single-number-i/
func SingleNumberFind(nums []int) int {
	frequency := make(map[int]int)

	for _, value := range nums {
		if _, ok := frequency[value]; !ok {
			frequency[value] = 1
		} else {
			frequency[value] += 1
		}
	}

	for _, value := range frequency {
		if value == 1 {
			return value
		}
	}
	return 0
}

// https://leetcode.com/problems/single-number-iii/
func SingleNumberFilter(nums []int) []int {
	frequency := make(map[int]int)
	var results []int
	for _, value := range nums {
		if _, ok := frequency[value]; !ok {
			frequency[value] = 1
		} else {
			frequency[value] += 1
		}
	}

	for key, value := range frequency {
		if value == 1 {
			results = append(results, key)
		}
	}
	return results
}

File: rust/src/problems/single_number.rs
use std::collections::HashMap;

// https://leetcode.com/problems/single-number/
// https://leetcode.com/problems/single-number-ii/
// more generic to cater to more than twice
pub fn find(nums: Vec<i32>) -> i32 {
  return nums
    .into_iter()
    .fold(HashMap::new(), |mut acc: HashMap<i32, i32>, curr| {
      *acc.entry(curr).or_insert(0) += 1;
      return acc;
    })
    .into_iter()
    .find_map(|(n, f)| -> Option<i32> {
      return if f == 1 { Some(n) } else { None };
    })
    .unwrap();
}

// https://leetcode.com/problems/single-number-iii/
// https://stackoverflow.com/questions/28909583/removing-entries-from-a-hashmap-based-on-value
pub fn filter(nums: Vec<i32>) -> Vec<i32> {
  return nums
    .into_iter()
    .fold(HashMap::new(), |mut acc: HashMap<i32, i32>, curr| {
      *acc.entry(curr).or_insert(0) += 1;
      return acc;
    })
    .into_iter()
    .filter_map(|(n, f)| -> Option<i32> {
      return if f == 1 { Some(n) } else { None };
    })
    .collect::<Vec<_>>();
}
